src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs

```diff
@@ -0,0 +1,269 @@
+        {
+            var references = await SymbolFinder.FindReferencesAsync(symbol, solution, cancellationToken).ConfigureAwait(false);
+
+            // Exclude the following kind of symbols:
+            //  (a) Implicitly declared symbols (such as implicit fields backing properties)
+            //  (b) Symbols that can't be referenced by name (such as property getters and setters).
+            //  (c) Metadata only symbols, i.e. symbols with no location in source.
+            IEnumerable<ReferencedSymbol> filteredReferences = from reference in references
+                                                               let symbolDef = reference.Definition
+                                                               where FilteringHelpers.FilterReference(symbolDef, reference)
+                                                               select reference;
+
+            return filteredReferences.Select(reference => Tuple.Create(reference, IsQueriedDefinition(reference, symbol)));
+        }
+
+        private static async Task<Tuple<SyntaxNode, IEnumerable<Tuple<ReferencedSymbol, bool>>>> FindReferencesAsync(Solution solution, Document document, SyntaxNode syntaxNode, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            var symbol = semanticModel.GetDeclaredSymbol(syntaxNode, cancellationToken);
+            if (symbol == null)
+            {
+                return null;
+            }
+
+            var references = await SymbolFinder.FindReferencesAsync(symbol, solution, cancellationToken).ConfigureAwait(false);
+
+            // Exclude the following kind of symbols:
+            //  (a) Implicitly declared symbols (such as implicit fields backing properties)
+            //  (b) Symbols that can't be referenced by name (such as property getters and setters).
+            //  (c) Metadata only symbols, i.e. symbols with no location in source.
+            IEnumerable<ReferencedSymbol> filteredReferences = from reference in references
+                                                               let symbolDef = reference.Definition
+                                                               where FilteringHelpers.FilterReference(symbolDef, reference)
+                                                               select reference;
+
+            var allReferences =
+                filteredReferences.Select(reference => Tuple.Create(reference, IsQueriedDefinition(reference, symbol)));
+
+            // Search through all other projects to see if we can a matching symbol in other projects
+            foreach (var additionalProject in document.Project.Solution.Projects)
+            {
+                if (additionalProject == document.Project)
+                {
+                    continue;
+                }
+
+                var additionalDocument = additionalProject.Documents.FirstOrDefault(d => document.FilePath.Equals(d.FilePath, StringComparison.OrdinalIgnoreCase));
```

Comment:
this lambda seems can be created once outside of loop and reused since document doesnt change.